### PR TITLE
Closes #52 | Add cosem dataloader

### LINKED ISSUE #52: Create dataset loader for Corpus of Singapore English Messages (CoSEM)
Dataloader name: `cosem/cosem.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?cosem

|     Dataset| cosem |
|-------------|---|
| Description | The CoSEM dataset consists of over 900,000 lines of online messages from the messaging platform WhatsApp collected from personal chat logs of students enrolled in an advanced sociolinguistics class from the National University of Singapore. Messages collected were from 2016 to 2019.  The dataset is in .txt format, where each line of utterance is tagged with a unique identifier that includes its metadata such as line number, year message was sent, and age and nationality of sender. |
| Subsets | - |
| Languages | eng |
| Tasks | Language Modeling |
| License | Creative Commons Zero v1.0 Universal (cc0-1.0) |
| Homepage | https://github.com/wdwgonzales/CoSEM/blob/main/Corpus/COSEM_v4_publicrelease_SEP172023.zip |
| HF URL | - |
| Paper URL | https://onlinelibrary.wiley.com/doi/full/10.1111/weng.12534 |

### PR DESCRIPTION
**Pull request by Enliven26**
Closes [#52](https://github.com/SEACrowd/seacrowd-datahub/issues/52)

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: seacrowd/sea_datasets/cosem/cosem.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+_SUPPORTED_TASKS = [Tasks.SELF_SUPERVISED_PRETRAINING]
+_SUPPORTED_SCHEMA_STRINGS = [f"seacrowd_{str(TASK_TO_SCHEMA[task]).lower()}" for task in _SUPPORTED_TASKS]
+
+_SOURCE_VERSION = "1.0.0"
+
+_SEACROWD_VERSION = "1.0.0"
+
+
+class MyWsl2023(datasets.GeneratorBasedBuilder):
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/cosem/cosem.py:115]: Maybe use CoSEM instead?

```suggestion
class CoSEMDataset(datasets.GeneratorBasedBuilder):
```

**Review by ljvmiranda921 — COMMENTED** [review]
Hi @Enliven26 , thanks so much for the PR! I have a few nits, but the data loader works as expected 👍

**Comment by Enliven26** [comment]
Hi, please review my recent changes, Thanks! @ljvmiranda921

**Review by ljvmiranda921 — APPROVED** [review]
LGTM!

**Comment by holylovenia** [comment]
A friendly reminder for @khelli07 to review.

**Comment by khelli07** [comment]
Hi, sorry for the late review. Before I get into my review, let me confirm first, you are taking the data per line, right? Then, divide the data into two based on `"\t"` character.
image

### If yes, have you consider these cases?
However, although this covers most cases, I found out that there is at least one edge case. You can see in the `COSEM_v4_chunk_1016.txt`, that the text is not in one row.
image

### How did I find this?
Basically, because you used `id` with `counter`, I wonder what would happen if we use the lefthand-side ID, e.g. `COSEM:19CM13-2652-21CHMIC-2017` because I think this can also act as a unique identifier. When I tried to change the ID to the `blocks[0]`, turned out that a lot of the IDs were empty. Which is why I suspected something was wrong. So, with this approach, you will have many both (or either) empty ID in `blocks[0]` and empty text in the `blocks[1]`.

### My suggestion:
You can use the id, e.g. `<COSEM:19CM13-2652-21CHMIC-2017>` for the identifier as well as the separator. When first loading the text, you can add a sentinel token at the EOF, e.g. `<COSEM:XXX>`. So if the text chunk, for example, is
```
<COSEM:ID-1> TEXT <TEXT>
TEXT

TEXT
<COSEM:ID-2> TEXT 

<COSEM:XXX> // EOF
```
You can find a pattern that the *TEXT* is between the COSEM:ID's `>` and the next COSEM:ID's `<`. 

That's what I thought though, feel free to use other ways. But, keep in mind that there are cases where the `<` and `>` are in the text, e.g.
image

which is why you can not just use them as a "separator".

<hr>

Oops, sorry, too long. Haha! cc @holylovenia @ljvmiranda921 (I would also like feedback from you guys on this, thanks)

**Comment by Enliven26** [comment]
> Hi, sorry for the late review. Before I get into my review, let me confirm first, you are taking the data per line, right? Then, divide the data into two based on `"\t"` character. image
> 
> ### If yes, have you consider these cases?
> Although this covers most cases, I doubt that it also covers edge cases. You can see in the `COSEM_v4_chunk_1016.txt`, that the text is not in one row. image
> 
> ### How did I find this?
> Basically, because you used `id` with `counter`, I wonder what would happen if we use the lefthand-side ID, e.g. `COSEM:19CM13-2652-21CHMIC-2017` because I think this can also act as a unique identifier. When I tried to change the ID to the `blocks[0]`, turned out that a lot of the IDs were empty. Which is why I suspected something was wrong. So, with this approach, you will have many both (or either) empty ID in `blocks[0]` and empty text in the `blocks[1]`.
> 
> ### My suggestion:
> You can use the id, e.g. `<COSEM:19CM13-2652-21CHMIC-2017>` for the identifier as well as the separator. When first loading the text, you can add a sentinel token at the EOF, e.g. `<COSEM:XXX>`. So if the text chunk, for example, is
> 
> ```
> <COSEM:ID-1> TEXT <TEXT>
> TEXT
> 
> TEXT
> <COSEM:ID-2> TEXT 
> 
> <COSEM:XXX> // EOF
> ```
> 
> You can find a pattern that the _TEXT_ is between the COSEM:ID's `>` and the next COSEM:ID's `<`.
> 
> That's what I thought though, feel free to use other ways. But, keep in mind that there are cases where the `<` and `>` are in the text, e.g. image
> 
> which is why you can not just use `<>` as a "separator".
> 
> Oops, sorry, too long. Haha! cc @holylovenia @ljvmiranda921 (I would also like feedback from you guys on this, thanks)

Hi @khelli07 , please review my recent changes, thanks!

**Comment by khelli07** [comment]
Hi, thank you for the fix! I'll review it by this week.

**Comment by khelli07** [comment]
Hi, I still find the bug 
image

Just for a note, your result is missing the "xt" in "Context" and "iting" in "Playwr"

The expected result should be like this
image

**Comment by khelli07** [comment]
I'll attach my code for reference. You can give me feedback too!
```
    def _generate_examples(self, path: str) -> Tuple[int, Dict]:
        """Yields examples as (key, example) tuples."""

        files = os.listdir(path)
        file_paths = [os.path.join(path, file) for file in files]
        pattern = r"<(COSEM:.*?)>(.*?)(?=<COSEM:|$)"

        s = {}
        for file_path in file_paths:
            with open(file_path, "r", encoding="utf-8") as file:
                text = file.read()
                maxchar = max(maxchar, len(text))

                matches = re.findall(pattern, text, re.DOTALL)
                for match in matches:
                    key = match[0].strip()
                    value = match[1].strip()
                    if key in s:
                        continue
                    s[key] = value

                    yield key, {"id": key, "text": value}

**Comment by Enliven26** [comment]
> I'll attach my code for reference. You can give me feedback too!
> 
> ```
>     def _generate_examples(self, path: str) -> Tuple[int, Dict]:
>         """Yields examples as (key, example) tuples."""
> 
>         files = os.listdir(path)
>         file_paths = [os.path.join(path, file) for file in files]
>         pattern = r"<(COSEM:.*?)>(.*?)(?=<COSEM:|$)"
> 
>         s = {}
>         for file_path in file_paths:
>             with open(file_path, "r", encoding="utf-8") as file:
>                 text = file.read()
>                 maxchar = max(maxchar, len(text))
> 
>                 matches = re.findall(pattern, text, re.DOTALL)
>                 for match in matches:
>                     key = match[0].strip()
>                     value = match[1].strip()
>                     if key in s:
>                         continue
>                     s[key] = value
> 
>                     yield key, {"id": key, "text": value}
> ```



> I'll attach my code for reference. You can give me feedback too!
> 
> ```
>     def _generate_examples(self, path: str) -> Tuple[int, Dict]:
>         """Yields examples as (key, example) tuples."""
> 
>         files = os.listdir(path)
>         file_paths = [os.path.join(path, file) for file in files]
>         pattern = r"<(COSEM:.*?)>(.*?)(?=<COSEM:|$)"
> 
>         s = {}
>         for file_path in file_paths:
>             with open(file_path, "r", encoding="utf-8") as file:
>                 text = file.read()
>                 maxchar = max(maxchar, len(text))
> 
>                 matches = re.findall(pattern, text, re.DOTALL)
>                 for match in matches:
>                     key = match[0].strip()
>                     value = match[1].strip()
>                     if key in s:
>                         continue
>                     s[key] = value
> 
>                     yield key, {"id": key, "text": value}
> ```

Hi, I think your solution is better. I didn't think to get all text in the file instead of read per line before. However, I remove the duplicate id checking since I assume the dataset must have unique ids. Please review my recent changes, thank! @khelli07

**Comment by khelli07** [comment]
Hey, I do think it is best to include the duplicate id checking since our assumption might not always be correct! :)

**Comment by Enliven26** [comment]
> if key in s:
>                         continue
>                     s[key] = value

done!

**Comment by Enliven26** [comment]
> Hey, I do think it is best to include the duplicate id checking since our assumption might not always be correct! :)

done!

**Review by khelli07 — APPROVED** [review]
LGTM! Thank you Johanes!
Just need @ljvmiranda921 's review after the changes

**Review by ljvmiranda921 — APPROVED** [review]
LGTM! Merging in a few

**Pull request merged**